engine/engine.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package engine
+
+import (
+	"os"
+
+	"github.com/rs/zerolog"
+)
+
+type Engine struct {
+	log        zerolog.Logger
+	components []*Component
+
+	sig    chan os.Signal
+	done   chan struct{}
+	failed chan struct{}
+}
+
+// New creates a new engine.
+func New(log zerolog.Logger, name string, sig chan os.Signal) *Engine {
+	e := Engine{
+		log: log.With().Str("engine", name).Logger(),
+		sig: sig,
+	}
+
+	return &e
+}
+
+// Component registers a new component for the engine. Components will be shut down
+// in the same order as the one in which they were registered.
+func (e *Engine) Component(name string, run func() error, stop func()) *Engine {
+	c := Component{
+		log:  e.log.With().Str("component", name).Logger(),
+		run:  run,
+		stop: stop,
+	}
+
+	e.components = append(e.components, &c)
+
+	return e
+}
+
+// Run launches the engine components and waits for them to either finish successfully,
+// fail, or for an external signal to shut the engine down.
+func (e *Engine) Run() *Engine {
+	e.done = make(chan struct{}, len(e.components))
+	e.failed = make(chan struct{}, len(e.components))
+
+	for _, component := range e.components {
+		go component.Run(e.done, e.failed)
+	}
+
+	// Here, we are waiting for a signal, or for one of the components to fail
+	// or finish. In both cases, we proceed to shut down everything, while also
+	// entering a goroutine that allows us to force shut down by sending
+	// another signal.
+	select {
+	case <-e.sig:
+		e.log.Info().Msg("engine stopping")
```

Comment:
This is where the engine's `Stop()` method should be invoked, no?

Potentially, components `stop` function could also be allowed to return an error.
Of the components I saw, only the metrics server shutdown **could** return an error (shutdown returns one which we do not propagate). Still, it may not be bad to allow the component shutdown code the be able to communicate an error.

The engine's `Stop()` method would need to be able to handle partial/multiple component shutdown failure (perhaps return a multierror)?

I don't think it's some huge benefit though.

Reply:
But we're still not trying to stop the engine, no?
(I think this is the most railroad-y sentence I ever wrote on Github)

Reply:
Reading it again - we're not really calling the `engine.Stop()` method anywhere outside of the package. It can then be unexported, if it's a thing the engine does internally.

Reply:
I initially left it exported because there could be cases where you want to create an engine and then stop it programmatically rather than depend on signals/it stopping by itself; but you're right it's not used currently so no need to export it for now.